### PR TITLE
Adding ticking functionality for block polling

### PR DESCRIPTION
**Pull request by paweljakubas**
# Issue Number

#3 

# Overview

- [x] I have added bracket with ticking function that takes block logic, action taken when blocks are received and tick time
- [x] I added tests that tests ticking function

# Comments

### CODE REVIEW

File: src/Cardano/Wallet/BlockSyncer.hs [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+
+
+newtype BlocksConsumed = BlocksConsumed [(Hash "BlockHeader")] deriving (Show, Eq)
+
+data BlockLayerMockup = BlockLayerMockup {
+      getBlock      :: (Hash "BlockHeader") -> IO Block
+    , getEpoch      :: Word64 -> IO [Block]
+    , getNetworkTip :: IO ((Hash "BlockHeader"),BlockHeader)
+    }
```
Comment (KtorZ) [inline src/Cardano/Wallet/BlockSyncer.hs]: This, and the implementation below of the mockup are only for testing I presume :thinking: ? I wouldn't expect this to be needed for writing and testing the ticking function. Cf next comment.

File: src/Cardano/Wallet/BlockSyncer.hs [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+runBlockLayerMockup
+    :: Int
+    -> [((Hash "BlockHeader"),Block)]
+    -> Int
+    -> IO BlockLayerMockup
```
Comment (KtorZ) [inline src/Cardano/Wallet/BlockSyncer.hs]: General remark about function API, but primitive types like `Int` or `String` communicate badly about their usage. It's always a good idea to add a comment explaining what the parameter is for. It's usually an even better idea to create a proper type to encapsulate that semantic.

File: src/Cardano/Wallet/BlockSyncer.hs [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+blockSyncerBracket
+    :: IORef BlocksConsumed
+    -> BlockLayerMockup
+    -> Int
+    -> IO ()
+blockSyncerBracket blocksPersisted blockLayer period =
+    bracket
+      (return ()) --(putStrLn "setting up for ticking function")
+      (\_->return ()) --(\_ -> putStrLn "cleaning for ticking function")
```
Comment (KtorZ) [inline src/Cardano/Wallet/BlockSyncer.hs]: Commented-out code shouldn't be pushed / committed (unless it serves as an example for documentation), which isn't the case here :+1:

File: src/Cardano/Wallet/BlockSyncer.hs [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+                  [] -> pure [currentHeaderHash]
+                  lastConsumed : _ | lastConsumed == currentHeaderHash ->
+                                         pure headerHashesConsumed
+                                   | lastConsumed == previousHeaderHash ->
+                                         pure $ currentHeaderHash : headerHashesConsumed
+                                   | otherwise -> do
+                                         missingHashes <- retrieveMissingBlocks [previousHeaderHash]
+                                                          headerHashesConsumed
+                                                          blockLayer
+                                         pure $ currentHeaderHash : (missingHashes ++ headerHashesConsumed)
```
Comment (KtorZ) [inline src/Cardano/Wallet/BlockSyncer.hs]: Make sure to follow the coding standards on variable-length indentation :pray:

File: src/Cardano/Wallet/BlockSyncer.hs [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    bracket
+      (return ()) --(putStrLn "setting up for ticking function")
+      (\_->return ()) --(\_ -> putStrLn "cleaning for ticking function")
+      (tickingFunction blocksPersisted blockLayer period)
+
+tickingFunction
+    :: IORef BlocksConsumed
+    -> BlockLayerMockup
+    -> Int
+    -> ()
```
Comment (KtorZ) [inline src/Cardano/Wallet/BlockSyncer.hs]: This is very dubious :no_mouth:

File: src/Cardano/Wallet/BlockSyncer.hs [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+                                   | lastConsumed == previousHeaderHash ->
+                                         pure $ currentHeaderHash : headerHashesConsumed
+                                   | otherwise -> do
+                                         missingHashes <- retrieveMissingBlocks [previousHeaderHash]
+                                                          headerHashesConsumed
+                                                          blockLayer
+                                         pure $ currentHeaderHash : (missingHashes ++ headerHashesConsumed)
+          writeIORef initialState $ BlocksConsumed updatedHeaderHashes
+          threadDelay (delayPeriod*1000*1000)
+          go previousState
```
Comment (KtorZ) [inline src/Cardano/Wallet/BlockSyncer.hs]: Okay, so here's the thing. This function seems a bit too explicit and conflates several responsabilities into one. As a rule of thumb, function should do one and one thing, and try not to work at different abstraction level. If we step back a little and re-think a bit the requirements / design of such function. 

- It's about ticking / polling a source of blocks and, proceed with some action when a new block is identified.

This means that we need:

- A way to poll a source and get the latest block from a source
- An action to execute when a new block is received

We want to leave it to the caller to define such behavior, and the ticking function should just bother 
about keeping track of that latest block and calling an action when a new one is discovered.
How blocks are fetched and what action is executed matters not; So we can inject the corresponding
behaviors from the arguments:

```hs
tickingFunction 
    :: IO Block
    -> (Block -> IO ())
    -> IO ()
tickingFunction getNextBlock action = do
  -- ...
```

Then, the implementation shouldn't do much more than keeping track of the
previous block fetched, and regularly polling the chain to fetch the one,
calling the action whenever the block fetched is different from the latest
block fetched. 

There's no particular need for mocking here, because the behavior of how you get blocks doesn't matter 
for the ticking function; this incidentally makes testing way easier because you get to choose that
behavior in your testing code!

File: src/Cardano/Wallet/BlockSyncer.hs [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+
+retrieveMissingBlocks
+    :: [(Hash "BlockHeader")]
+    -> [(Hash "BlockHeader")]
+    -> BlockLayerMockup
+    -> IO [(Hash "BlockHeader")]
+retrieveMissingBlocks missing headerHashesConsumed layer = do
```
Comment (KtorZ) [inline src/Cardano/Wallet/BlockSyncer.hs]: This is out of scope for this task.

File: test/unit/Cardano/Wallet/BlockSyncerSpec.hs [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+import Test.QuickCheck
+    ( generate, vector )
+import Test.QuickCheck.Gen
+    ( Gen )
+
+
+import Cardano.Wallet.BlockSyncer
+    ( BlocksConsumed (..), blockSyncerBracket, runBlockLayerMockup )
+import Cardano.Wallet.Primitive
+    ( Block (..), BlockHeader (..), Hash (..) )
```
Comment (KtorZ) [inline test/unit/Cardano/Wallet/BlockSyncerSpec.hs]: Make sure to follow the coding standards for imports :pray:  (two groups, qualified vs explicit imports. We do not make any more groups based on whether the code comes from our code base or not).

File: test/unit/Cardano/Wallet/BlockSyncerSpec.hs [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+        it "when the mockup ticking is 5x more frequent" $ do
+            testBlocksPolling 10 5 1
+        it "when the mockup ticking is 2x more frequent" $ do
+            testBlocksPolling 10 2 1
+
+    where
+        testBlocksPolling
+            :: Int
+            -> Int
+            -> Int
```
Comment (KtorZ) [inline test/unit/Cardano/Wallet/BlockSyncerSpec.hs]: In such case, think about using a record type to make the intent clear about what the parameters are.

File: test/unit/Cardano/Wallet/BlockSyncerSpec.hs [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+        testBlocksPolling numberOfBlocks mockupPeriod blockTickerPeriod = do
+            consecutiveBlocks <- liftIO $ mkConsecutiveTestBlocks numberOfBlocks
+            let testTimeEvalInSec =  numberOfBlocks*mockupPeriod + 2*blockTickerPeriod
+            blockLayerMockup <- runBlockLayerMockup mockupPeriod consecutiveBlocks testTimeEvalInSec
+            initialHeaderData <- newIORef $ BlocksConsumed []
+            threadID <- forkIO $ blockSyncerBracket initialHeaderData blockLayerMockup blockTickerPeriod
+            threadDelay (testTimeEvalInSec*1000*1000)
+            killThread threadID
+            (BlocksConsumed finalHeaderHashes) <- readIORef initialHeaderData
+            finalHeaderHashes `shouldBe` (map fst consecutiveBlocks)
```
Comment (KtorZ) [inline test/unit/Cardano/Wallet/BlockSyncerSpec.hs]: Also, this function would deserve to be split in multiple ones that each encapsulate a particular responsibility. It would make the organization of the code a bit clearer between how the test is being set up, what the test action is and what are the expectations.

**Review by KtorZ — CHANGES_REQUESTED** [review]
I think you were a bit off track here and went for more than what was needed. It's important to separate concerns in your implementation and make sure that code isn't needlessly coupled to some other code. 

It would have been nice to get a PR much sooner actually to prevent this; as per our new development process, we want to keep on at least one PR a day so make sure to break a task down to iterations that are achievable within a day. Mocking isn't usually a good idea, but inversion of control may help a lot. 

Also, make sure to go through and acknowledge the [coding standards](https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards) once again. It's important for us all to be consistent on the practices.

File: src/Cardano/Wallet/BlockSyncer.hs (hunk 1 above)
Comment (paweljakubas) [reply src/Cardano/Wallet/BlockSyncer.hs]: it was supposed to act like that. in the light of the refactoring it is not used anymore

File: src/Cardano/Wallet/BlockSyncer.hs (hunk 2 above)
Comment (paweljakubas) [reply src/Cardano/Wallet/BlockSyncer.hs]: :+1:

File: src/Cardano/Wallet/BlockSyncer.hs (hunk 3 above)
Comment (paweljakubas) [reply src/Cardano/Wallet/BlockSyncer.hs]: I put it like that for the sake of future logging.

File: src/Cardano/Wallet/BlockSyncer.hs (hunk 4 above)
Comment (paweljakubas) [reply src/Cardano/Wallet/BlockSyncer.hs]: :+1:

File: src/Cardano/Wallet/BlockSyncer.hs (hunk 6 above)
Comment (paweljakubas) [reply src/Cardano/Wallet/BlockSyncer.hs]: complied. Thanks for detailed directions

File: src/Cardano/Wallet/BlockSyncer.hs (hunk 7 above)
Comment (paweljakubas) [reply src/Cardano/Wallet/BlockSyncer.hs]: yes, I removed that

File: test/unit/Cardano/Wallet/BlockSyncerSpec.hs (hunk 8 above)
Comment (paweljakubas) [reply test/unit/Cardano/Wallet/BlockSyncerSpec.hs]: :+1:

File: test/unit/Cardano/Wallet/BlockSyncerSpec.hs (hunk 9 above)
Comment (paweljakubas) [reply test/unit/Cardano/Wallet/BlockSyncerSpec.hs]: :-)

File: test/unit/Cardano/Wallet/BlockSyncerSpec.hs (hunk 10 above)
Comment (paweljakubas) [reply test/unit/Cardano/Wallet/BlockSyncerSpec.hs]: I will do it in the final commit

File: src/Cardano/Wallet/BlockSyncer.hs (hunk 3 above)
Comment (KtorZ) [reply src/Cardano/Wallet/BlockSyncer.hs]: Don't :) 
If it's not needed, don't do it. We favor simple code and refactor when we need to add functionalities :+1:

File: test/unit/Cardano/Wallet/BlockSyncerSpec.hs [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+
+            blockToIOaction <- writeToIORefAction consumerData consecutiveBlocks
+            let blockDelivery = pushNextBlocks producerData
+            _ <- forkIO $ blockSyncerBracket blockDelivery blockToIOaction tickTime
+            threadDelay (evalTestTimeInSec*1000*1000)
+
+            (BlocksConsumed consumedData) <- readIORef consumerData
+
+            consumedData `shouldBe` ((map fst . reverse) consecutiveBlocks)
+
```
Comment (paweljakubas) [inline test/unit/Cardano/Wallet/BlockSyncerSpec.hs]: there is further refactoring under way

File: src/Cardano/Wallet/BlockSyncer.hs [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+    -> (Block -> IO ())
+    -- ^ action taken on a new block
+    -> Int
+    -- ^ tick time in seconds
+    -> IO ()
+blockSyncerBracket getNextBlocks action tickTime =
+    bracket
+      (return ())
+      (\_->return ())
+      (tickingFunction getNextBlocks action tickTime)
```
Comment (KtorZ) [inline src/Cardano/Wallet/BlockSyncer.hs]: So, this is the kind of "not needed yet abstractions" I am often referring too. The bracket doesn't buy us anything here, and there's no particular need to have such bracket. Maybe later we'll have to get some resources before calling the ticking function, but that's later. As for now, we just want to have a ticking function and test it.

File: src/Cardano/Wallet/BlockSyncer.hs [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+
+tickingFunction
+    :: IO [Block]
+    -> (Block -> IO ())
+    -> Int
```
Comment (KtorZ) [inline src/Cardano/Wallet/BlockSyncer.hs]: Would be better to get a better type here which convey a clear meaning -- or... not having any at all. The time can just be a hard-coded constant on top of the module, if we ever need to change that, we'll make it a parameter. But again, for now, this parameter isn't needed or required.

File: src/Cardano/Wallet/BlockSyncer.hs (hunk 5 above)
Comment (KtorZ) [reply src/Cardano/Wallet/BlockSyncer.hs]: I still wonder why have this unit argument here?

File: src/Cardano/Wallet/BlockSyncer.hs [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+    -> ()
+    -> IO ()
+tickingFunction getNextBlocks action tickTime _ =
+    go `catch`
+    (\(e :: SomeException) ->
+            let msg = "Terminating tickingFunction of BlockSyncer due to " % F.shown
+            in (putStrLn . showText) $ sformat msg e
+    )
```
Comment (KtorZ) [inline src/Cardano/Wallet/BlockSyncer.hs]: That is rather problematic to handle the error here in such way.. If the ticking function dies... as for now, we probably wants the wallet to die too. That's a fatal error and here, we'll just be showing a message in the console and silently fade away. The wallet server will keep running and... nothing will happen until we eventually notice that there's no block getting fetched anymore. 

So, until further needs, let just _do nothing_ and let the exception bubbles to the caller.

File: src/Cardano/Wallet/BlockSyncer.hs [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+    where
+      go :: IO ()
+      go = do
+          blocksToProcess <- getNextBlocks
+          mapM_ action blocksToProcess
```
Comment (KtorZ) [inline src/Cardano/Wallet/BlockSyncer.hs:58]: We miss a crucial part here which is: controlling that those blocks are new blocks. So, maybe the requirements weren't really clear but, here's the context:

- We have a networking layer that allows us to fetch the latest block from the network. It doesn't do much more than, fetching the tip; there's no particular logic involved apart from implementing the right communication protocol (e.g. an http request to the bridge). 

- Then, the ticking function contains an extra level of logic where, we want to know if we need to process the current block. Maybe we already have, and we do not want to process it. So, the action should only be called with yet unprocessed blocks.

File: test/unit/Cardano/Wallet/BlockSyncerSpec.hs [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+            producerData <- newIORef $ BlocksToInject (chunkSizesToTest,consecutiveBlocks)
+
+            blockToIOaction <- writeToIORefAction consumerData consecutiveBlocks
+            let blockDelivery = pushNextBlocks producerData
+            _ <- forkIO $ blockSyncerBracket blockDelivery blockToIOaction tickTime
+            threadDelay (evalTestTimeInSec*1000*1000)
+
+            (BlocksConsumed consumedData) <- readIORef consumerData
+
+            consumedData `shouldBe` ((map fst . reverse) consecutiveBlocks)
```
Comment (KtorZ) [inline test/unit/Cardano/Wallet/BlockSyncerSpec.hs]: It is rather unclear to me what the tests are testing here :thinking: ? A comment or a longer scenario would definitely helps. ASCII diagrams are also very welcome.

**Review by KtorZ — CHANGES_REQUESTED** [review]

File: src/Cardano/Wallet/BlockSyncer.hs (hunk 12 above)
Comment (paweljakubas) [reply src/Cardano/Wallet/BlockSyncer.hs]: removed bracket as requested

File: src/Cardano/Wallet/BlockSyncer.hs (hunk 13 above)
Comment (paweljakubas) [reply src/Cardano/Wallet/BlockSyncer.hs]: went with Second from `time-units`

File: src/Cardano/Wallet/BlockSyncer.hs [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+import Data.Time.Units
+    ( Second, toMicroseconds )
+
+import qualified Data.List as L
+
+
+newtype BlockHeadersConsumed = BlockHeadersConsumed [BlockHeader] deriving (Show, Eq)
+
+storingLimit :: Int
+storingLimit = 2140
```
Comment (KtorZ) [inline src/Cardano/Wallet/BlockSyncer.hs]: Did you mean `2160` :thinking: ?

File: src/Cardano/Wallet/BlockSyncer.hs (hunk 14 above)
Comment (paweljakubas) [reply src/Cardano/Wallet/BlockSyncer.hs]: removed and also formatting dependency

File: src/Cardano/Wallet/BlockSyncer.hs (hunk 15 above)
Comment (paweljakubas) [reply src/Cardano/Wallet/BlockSyncer.hs:58]: done, now tickingFunction is responsible not to take action on the block it previously consumed. I added the comment also specifying known limitation (lack of handling block gaps)

File: test/unit/Cardano/Wallet/BlockSyncerSpec.hs (hunk 16 above)
Comment (paweljakubas) [reply test/unit/Cardano/Wallet/BlockSyncerSpec.hs]: I wanted to check first **ExactlyOnce** delivery. When the ticking function gained not-consuming-the-same-block-guarrantee I added also **AtLeastOnce** delivery test

File: src/Cardano/Wallet/BlockSyncer.hs (hunk 17 above)
Comment (paweljakubas) [reply src/Cardano/Wallet/BlockSyncer.hs]: yeah, when we later tackle rollback security parameter comes here (I thought about it but made blunder here putting 2140 instead of 2160). For the time being I wanted to just add some upper bound just in order to not surpass memory limit.

File: cardano-wallet.cabal [hunk 18]
```diff
@@ -37,13 +37,16 @@ library
     , cborg
     , containers
     , deepseq
+    , text
```
Comment (paweljakubas) [inline cardano-wallet.cabal:47]: not needed - will remove in the final commit

**Comment by paweljakubas** [comment]
problems with tests are because of IORef use. I will refactor the code with MVar

File: src/Cardano/Wallet/BlockSyncer.hs [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+          go $ BlockHeadersConsumed
+              $ take storingLimit
+              $ map header blocksToProcess ++ headersConsumed
+
+      checkIfAlreadyConsumed
+          :: [BlockHeader]
+          -> Block
+          -> Bool
+      checkIfAlreadyConsumed consumedHeaders (Block theHeader _) =
+          theHeader `L.notElem` consumedHeaders
```
Comment (KtorZ) [inline src/Cardano/Wallet/BlockSyncer.hs:69]: Okay, this looks good now. If I may throw a naming suggestion:

- ~~BlockHeaderConsumed~~ ---> `Known` with a type parameter, BlockHeader is rather unnecessary here, it's already in the type: `Known [BlockHeader]`

- ~~checkIfAlreadyConsumed~~ ---> `isKnown`, less prose is good :) 

- ~~tickTime~~ --> `delay`, well, that's what it is in the end. 

- ~~blocksDownloaded~~ --> `blocks`, again, no need to be extra verbose about this. By keeping your function small, the context is enough to understand this. And, in the end, it doesn't really matter if blocks have been _downloaded_, or _fetched_ or _unserialized_ .. Unless it helps disambiguate things, go for a short name.

- ~~`blocksToProcess`~~ --> `block'` or `unknownBlock`, just to be aligned with the above suggestions

- ~~`consumedHeaders`~~ --> `knownHeaders`, also, to be aligned

- ~~`theHeader`~~ --> `h` is probably enough because the function is really short, or, do not pattern match and go for `header block` or `header b` 


Short names are usually good. Too short is sometimes not very helpful, though it make sense when working on abstractions (like in `filter :: (a -> Bool) -> [a] -> [a]`, one wouldn't write `filter :: (theElement -> Bool) -> theList -> theList`).

File: test/unit/Cardano/Wallet/BlockSyncerSpec.hs [hunk 20]
```diff
@@ -0,0 +1,177 @@
...
+generateBlockChunkSizes = do
+    numberOfTicks <- generate $ choose (1,15)
+    generate $ generateBlockChunks numberOfTicks
+        where
+            generateBlockChunks
+                :: Int
+                -> Gen [Int]
+            generateBlockChunks numberOfTicks = do
+                let chunkSizeGen = choose (0,15)
+                vectorOf numberOfTicks chunkSizeGen
```
Comment (KtorZ) [inline test/unit/Cardano/Wallet/BlockSyncerSpec.hs]: It would have been nice to have a chat about the approach on testing for this. I think you've got the idea, but the implementation of the tests has gone slightly too complex IMO. I'll make a suggestion a bit later but it would be nice to go through this together.

**Review by KtorZ — CHANGES_REQUESTED** [review]

File: src/Cardano/Wallet/BlockSyncer.hs [hunk 21]
```diff
@@ -0,0 +1,64 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+
+-- |
+-- Copyright: © 2018-2019 IOHK
+-- License: MIT
+--
+-- This module contains the ticking function that is responsible for invoking block acquisition functionality and executing it in periodic fashion.
+--
+-- Known limitations: ticking function makes sure action is not executed on already consumed block, but does not check and handle block gaps (aka catching up)
```
Comment (jonathanknowles) [inline src/Cardano/Wallet/BlockSyncer.hs]: Hi there! We'd need to wrap this to bring the above comment in line with https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#limit-line-length-to-80-characters

Perhaps like this:
```hs
-- This module contains the ticking function that is responsible for invoking
-- block acquisition functionality and executing it in periodic fashion.
--
-- Known limitations: the ticking function makes sure action is not executed on
-- already consumed block, but does not check and handle block gaps (aka
-- catching up).
```

File: src/Cardano/Wallet/BlockSyncer.hs [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+    ( Block (..), BlockHeader )
+import Control.Concurrent
+    ( threadDelay )
+import Data.Time.Units
+    ( Second, toMicroseconds )
+
+import qualified Data.List as L
+
+
+newtype BlockHeadersConsumed = BlockHeadersConsumed [BlockHeader] deriving (Show, Eq)
```
Comment (jonathanknowles) [inline src/Cardano/Wallet/BlockSyncer.hs]: https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#limit-line-length-to-80-characters

Propose:
```hs
newtype BlockHeadersConsumed =
    BlockHeadersConsumed [BlockHeader]
    deriving (Show, Eq)
```

File: src/Cardano/Wallet/BlockSyncer.hs (hunk 21 above)
Comment (paweljakubas) [reply src/Cardano/Wallet/BlockSyncer.hs]: :+1:

File: src/Cardano/Wallet/BlockSyncer.hs (hunk 22 above)
Comment (paweljakubas) [reply src/Cardano/Wallet/BlockSyncer.hs]: :+1:

**Comment by paweljakubas** [comment]
rebased and squashed

**Comment by paweljakubas** [comment]
it seems we have :

**Comment by paweljakubas** [comment]
it seems we have : 
```
5  {-# OPTIONS_GHC -fno-warn-unused-imports #-}                                                                                                                                               
6                                                                                                                                                                                             
7  module Cardano.Wallet.BlockSyncerSpec                                                                                                                                                      
8      ( spec                                                                                                                                                                                 
9      , groups                                                                                                                                                                               
10     , duplicateMaybe                                                                                                                                                                       
11     ) where 
``` 
and weeder still keeps complaining about **groups** and **duplicateMaybe** not used outside module

**Comment by KtorZ** [comment]
@paweljakubas Whoops. Sorry about the pragma, I typically do that when developing to iterate quickly and I clean up the imports before making a PR. This should be removed, and imports cleaned up as well :+1:

**Comment by paweljakubas** [comment]
ahh, I have already added proper **.weeder.yaml** being the product of:
```
[pawel@arch cardano-wallet]$ weeder . --yaml > .weeder.yaml
[pawel@arch cardano-wallet]$ cat .weeder.yaml
- package:
  - name: cardano-wallet
  - section:
    - name: test:unit
    - message:
      - name: Weeds exported
      - module:
        - name: Cardano.Wallet.BlockSyncerSpec
        - identifier:
          - duplicateMaybe
          - groups
```

We do not want it, right? Just remove pragma and superfluous exports?

**Review by KtorZ — APPROVED** [review]
